File: src/lib/nlp.ts
interface Category {
    id: string;
    name: string;
}

interface ParsedMessage {
    amount: number;
    description: string;
    type: "EXPENSE" | "INCOME";
    categoryId: string | null;
}

export function parseMessage(text: string, categories: Category[]): ParsedMessage {
    text = text.toLowerCase().trim();

    let amount = 0;
    let description = "Despesa WhatsApp";
    let type: "EXPENSE" | "INCOME" = "EXPENSE";
    let categoryId: string | null = null;

    // 1. Extract Amount (R$ 50, 50.00, 50,00, 50)
    const amountMatch = text.match(/R?\$?\s*(\d+([.,]\d{1,2})?)/);
    if (amountMatch) {
        amount = parseFloat(amountMatch[1].replace(',', '.'));
        text = text.replace(amountMatch[0], '').trim();
    }

    // 2. Determine Type
    if (['recebi', 'ganhei', 'renda', 'deposito', 'salario'].some(k => text.includes(k))) {
        type = "INCOME";
        description = "Receita WhatsApp";
    } else if (['gastei', 'paguei', 'compra', 'saque'].some(k => text.includes(k))) {
        type = "EXPENSE";
    }

    // 3. Clean Description
    const stopwords = ['gastei', 'paguei', 'recebi', 'ganhei', ' com ', ' no ', ' na ', ' em '];
    let cleanText = text;
    stopwords.forEach(word => {
        cleanText = cleanText.replace(new RegExp(word, 'g'), '');
    });

    if (cleanText.trim()) {
        description = cleanText.trim()
            .split(' ')
            .filter(w => w.length > 0)
            .map(w => w.charAt(0).toUpperCase() + w.slice(1))
            .join(' ');
    }

    // 4. Match Category
    for (const cat of categories) {
        if (description.toLowerCase().includes(cat.name.toLowerCase())) {
            categoryId = cat.id;
            break;
        }
    }

    // Fallback keyword matching
    if (!categoryId) {
        const keywords: Record<string, string[]> = {
            'food': ['comida', 'lanche', 'restaurante', 'ifood', 'mercado', 'pizza', 'almoço', 'janta'],
            'transport': ['uber', '99', 'taxi', 'bus', 'onibus', 'gasolina', 'posto', 'estacionamento'],
            'salary': ['salario', 'pagamento', 'freela', 'freelance'],
            'entertainment': ['cinema', 'jogo', 'steam', 'netflix', 'spotify'],
            'bills': ['luz', 'agua', 'internet', 'aluguel', 'conta']
        };

        for (const [catName, keys] of Object.entries(keywords)) {
            if (keys.some(k => description.toLowerCase().includes(k))) {
                const foundCat = categories.find(c => c.name.toLowerCase().includes(catName));
                if (foundCat) {
                    categoryId = foundCat.id;
                    break;
                }
            }
        }
    }

    return { amount, description, type, categoryId };
}

File: core/services/nlp.py
import re
from decimal import Decimal
from ..models import Category

def parse_message(text):
    """
    Parses a natural language message into transaction data.
    Enhanced to find Categories.
    """
    text = text.lower()
    
    # 1. Extract Amount
    # Matches: "50", "R$ 50", "50.00", "50,00"
    amount_match = re.search(r'R?\$?\s?(\d+[,.]?\d*)', text)
    amount = Decimal('0.00')
    if amount_match:
        amount_str = amount_match.group(1).replace(',', '.')
        amount = Decimal(amount_str)

    is_expense = True # Default
    if any(word in text for word in ['recebi', 'ganhei', 'salário', 'depósito', 'pix recebido']):
        is_expense = False

    # 3. Extract Description (Everything else)
    # Remove the amount from the text to leave the description
    description = text
    if amount_match:
        description = text.replace(amount_match.group(0), '').strip()
    
    # Clean up common connector words
    for word in ['gastei', 'no', 'na', 'com', 'em', 'de', 'do', 'da']:
        description = re.sub(r'\b' + word + r'\b', '', description).strip()

    # 4. Extract Category (Keyword matching)
    category = None
    all_categories = Category.objects.all()
    
    # Map keywords to categories
    # ideally this should be a DB field 'keywords' but for now we hardcode mappin
    keywords_map = {
        'comida': 'Alimentação', 'lanche': 'Alimentação', 'ifood': 'Alimentação', 'restaurante': 'Alimentação', 'mercado': 'Alimentação',
        'uber': 'Transporte', 'gasolina': 'Transporte', 'ônibus': 'Transporte', 'transporte': 'Transporte',
        'aluguel': 'Moradia', 'luz': 'Moradia', 'internet': 'Moradia', 'casa': 'Moradia',
        'remedio': 'Saúde', 'farmacia': 'Saúde', 'medico': 'Saúde',
        'cinema': 'Lazer', 'jogo': 'Lazer', 'sair': 'Lazer',
        'salario': 'Salário', 'pagamento': 'Salário',
        'freela': 'Freelance', 'projeto': 'Freelance',
    }
    
    # Check description against map
    found_cat_name = 'Outros' # Fallback
    for word, cat_name in keywords_map.items():
        if word in text:
            found_cat_name = cat_name
            break
            
    # Try to find the category object
    category = all_categories.filter(name__iexact=found_cat_name).first()
    if not category and is_expense:
         category = all_categories.filter(name='Compras').first() # Final fallback

    return {
        'amount': amount,
        'description': description.title() or "Nova Transação",
        'type': 'EXPENSE' if is_expense else 'INCOME',
        'category': category
    }
